### PR TITLE
Use global `Iterator.prototype` for down-level generators

### LINKED ISSUE #59513: Generators compiled with `--target es5 --lib esnext` do not support iterator helpers
### 🔎 Search Terms

generator iterator helper map esnext prototype

### 🕗 Version & Regression Information

- This changed in commit or PR #58222

### ⏯ Playground Link

https://www.typescriptlang.org/play/?target=1&lib=lib.esnext.d.ts&ts=5.6.0-beta#code/PTAEAEBcEMCcHMCmkBcpEGcCsAoEEAbASwCM1MA7RAD0hxwGMB7CjSUOUAXlAG0BGADSgATMIDMwgCwBdANw44AOgBu0AgFdMACgCUSgLbQADtqLcAfKHMAqUbrmh8vMaCnCAbMIAcM+gDMNCgZIIhZQO389UABvHFBQAE8iRAIAEwi+IVEJaXkcAF8cKP0jU3MuK1t7R3xYINCDRHRYWCZYNAByKKVe0pNO6wxQCiZ2aFBA4NCWIA

### 💻 Code

```ts
// @target: es5
// @lib: esnext

const ar = [1, 2, 3, 4];
ar.values().map(i => i * 2); // [2, 4, 6, 8]

function * f() {
  yield * [1, 2, 3, 4];
}
f().map(i => i * 2); // error: 'f(...).map' is not a function
```

### 🙁 Actual behavior

The last line above throws even when `Iterator.prototype` is present, be it from the host or a polyfill.

### 🙂 Expected behavior

The last line above should not throw when the global `Iterator.prototype` is present and can be used as a prototype.

### Additional information about the issue

_No response_

### PR DESCRIPTION
**Pull request by rbuckton**
If you are compiling generators with `--target ES5 --lib esnext` the types of those generators will also include the iterator helpers methods like `map`, `filter`, etc., though our down-level generator emit will not have those methods even when running with newer editions of V8 that do support them:

```ts
// @target: es5
// @lib: esnext

const ar = [1, 2, 3, 4];
ar.values().map(i => i * 2); // [2, 4, 6, 8]

function * f() {
  yield * [1, 2, 3, 4];
}
f().map(i => i * 2); // error: 'f(...).map' is not a function
```

Even when not running in latest V8, iterator helpers can be polyfilled, e.g.:
```js
if (typeof Iterator === "undefined") {
    let Iterator = function () {};
    Iterator.prototype = Object.getPrototypeOf(Object.getPrototypeOf([].values()));
    Iterator.prototype.map = ...;
    Iterator.prototype.filter = ...;
    globalThis.Iterator = Iterator;
}
```

To better support both scenarios, this makes a small change to our `__generator` helper to use the global `Iterator.prototype` object as the prototype for down-level generators, if it is present.

This also does the same for our `__asyncGenerator` helper when a global `AsyncIterator.prototype` object is present, even though async iterator helpers are still at Stage 2. Support for `AsyncIterator.prototype` seems harmless enough to add at this time but can be postponed to a later PR if necessary.

Please note that this will also require the same change in `tslib`.

Fixes #59513

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot: pack this

File: src/compiler/factory/emitHelpers.ts [hunk 1]
```diff
@@ -1102,8 +1102,8 @@ const generatorHelper: UnscopedEmitHelper = {
     priority: 6,
     text: `
             var __generator = (this && this.__generator) || function (thisArg, body) {
-                var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
-                return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
+                var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
```
Comment (jakebailey) [inline src/compiler/factory/emitHelpers.ts:1105]: I assume it's not a problem if someone redefines the global `Iterator`, given we don't care if someone does that for other global symbols?

**Review by jakebailey — APPROVED** [review]
This needs a `tslib` update too, right?

File: src/compiler/factory/emitHelpers.ts (hunk 1 above)
Comment (rbuckton) [reply src/compiler/factory/emitHelpers.ts:1105]: Correct

**Comment by rbuckton** [comment]
> This needs a `tslib` update too, right?

Correct. The tslib PR is https://github.com/microsoft/tslib/pull/267

**Pull request merged**